proposals/draft/19-layered-model-of-singer

```diff
@@ -0,0 +1,137 @@
+
+#### Layer 2: Standards
+The pieces that systems using Singer can rely on such as catalogs and discovery mode. These pieces required to implement to be considered as up to modern standards for Singer. Not implementing these is not an incompatibility, but all Singer actors should strive to implement the standards. The standards are more focused upon Singer actors themselves themselves (e.g., taps, targets), as opposed to the over-the-wire protocol at the heart of Singer. Some tooling that rely on these standards may not function if an actor does not implement it.
+
+Conceptually, this includes things like Command-Line Arguments, Catalog, Metadata Keys/Custom Metadata, Standard State Keys, etc.
+
+Some principles here:
+- Language agnostic, implementation specific, and generic regardless of use case
+- They help standardize the nitty-gritty to make writing frameworks and libraries easier
+
+#### Layer 3: Best Practices
+
+This is the set of things that Singer practicioners have figured out work in their space, but aren't sure whether they are globally generic or not yet. These best practices are generally not required, but are things that have been used as ways to make it easier to build taps.
+
+Some principles here:
+- Language agnostic, implementation specific, and use case specific.
+- These items are candidates for lower levels of the ecosystem, but maybe are not quite yet standards (i.e., are not at the level that they can be considered required)
+- They strive to make code more portable, readable, and usable for users and devs alike in a structural sense
+
+### Layer 4: Libraries and Frameworks
+This is where we get into the language specific stuff. Libraries like `singer-python` and/or `singer-clojure` or frameworks like the MeltanoSDK take the standards plus best practices and encode them in a way that makes sense for the patterns of each language. This is also a good place to be a test bed for things that might become standards.
+
+Principles:
+- Language specific
+- Generic use cases
+- These influence the way that code is written for their specific language
+
+### Layer 5: Tooling/Orchestration/UX/Infrastructure
+This is the bread and butter of organizations that adopt singer. As far as standards go, this is not generally something that will be voted upon by the Singer communicty as a whole. Instead, this is the state that all proprietary code falls in and the specific product offerings of practitioners fit. Nothing in this space is expected to be open source, but the option is there for things that may qualify for best practices.
+
+This layer can be considered an analog to the "Application Layer" of the OSI layered model of networking.
+
+Principles:
+- Specific to a particular vertical or market space
+- Specific to a single ops approach for deployment or other closer-to-the-metal conecerns
+- Not generalizable without either losing value in the source product
```

Comment:
This sentence I'm a little lost on but I get the general idea. Maybe

Not generalizable without losing value in the source product

Or just

Not generalizable


🤷 , need to work on my writing more anyways so you probably have a better idea (overall this writeup is great and easy to understand)

Reply:
Ah, yup heh. I think I wanted to keep some space for closed source pieces to be at least categorized here, but The intro probably explains that well enough. I like just "Not generalizable".